File: dialogio/make_stream.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

static void make_stream (const String *path, __UNUSED const String *mask, tSTREAM *stream)
{
  String searchpath[Pascal_FileNameLen+1];
  SearchRec_t search;
  uint16_t a = 0, b = 0; // count1, count2

  DBG_ENTER ("make_stream");

#if WIN32 || WIN64 || WINNT
  GetLogicalDriveStrings (sizeof (drive_list), drive_list);
#endif // WIN32 || WIN64 || WINNT

  for (char drive = 'A'; drive <= 'Z'; drive++)
#if GO32
    if (valid_drive (drive))
#else // !GO32
    if (valid_drive (drive, (String *)&stream->stuff[a].info))
#endif // !GO32
    {
      SetLength (stream->stuff[a].name, 1);
      GetStr (stream->stuff[a].name)[0] = drive;
      stream->stuff[a].attr = Pascal_VolumeId;
      stream->stuff[a].size = 0;
      a++;
    }

  // TODO: make a separate flag for this
  SetLength (stream->stuff[a].name, 3);
  GetStr (stream->stuff[a].name)[0] = '~';
  GetStr (stream->stuff[a].name)[1] = charmap.nbsp; // space
  GetStr (stream->stuff[a].name)[2] = '~';
  stream->stuff[a].attr = Pascal_VolumeId;
  a++;

  stream->drive_count = a;

#if GO32
  if (Pascal_DiskSize (UpCase (GetStr (path)[0]) - 'A' + 1) > 0)
  {
#endif // GO32
    CopyString (searchpath, path, Pascal_FileNameLen);
    AppendString (searchpath, WILDCARD_ASTERISK, Pascal_FileNameLen);

    Pascal_FindFirst (searchpath, Pascal_AnyFile - Pascal_VolumeId, &search);

    while ((Pascal_DosError == 0) && (a < MAX_FILES))
    {
      if (!(((search.Attr & Pascal_Directory) != 0) && (CompareStrings (search.Name, ONE_DOT_STR) == 0)))
        if (   ((search.Attr & Pascal_Directory) != 0)
            && !(   (CompareStrings (search.Name, TWO_DOTS_STR) == 0)
#if linux || unix
                 && (Length (path) == 1))) // `<PATHSEP>'
#else // !(linux || unix)
                 && (Length (path) == 3))) // `<disk>:<PATHSEP>'
#endif // !(linux || unix)
        {
          if (CompareStrings (search.Name, TWO_DOTS_STR) == 0)
            CopyString (search.Name, updir_str, Pascal_FileNameLen);

          CopyString (stream->stuff[a].name, search.Name, FILENAME_SIZE);
          stream->stuff[a].attr = search.Attr;
          a++;
        }

      Pascal_FindNext (&search);
    }

#if !GO32
    if ((Length (path) > 3) && (a == stream->drive_count))
    {
      CopyString (stream->stuff[a].name, updir_str, FILENAME_SIZE);
      stream->stuff[a].attr = search.Attr;
      a++;
    }
#endif // !GO32

    Pascal_FindFirst (searchpath, Pascal_AnyFile - Pascal_VolumeId - Pascal_Directory, &search);
    while ((Pascal_DosError == 0) && (a + b < MAX_FILES))
    {
      if (LookUpMask (search.Name))
      {
        String_t s = Lower_filename (search.Name);
        CopyString (search.Name, (String *)&s, Pascal_FileNameLen);
        CopyString (stream->stuff[a + b].name, search.Name, FILENAME_SIZE);
        stream->stuff[a + b].attr = search.Attr;
        stream->stuff[a + b].size = search.Size;
        b++;
      }

      Pascal_FindNext (&search);
    }
#if GO32
  }
#endif // GO32

  QuickSort (stream, stream->drive_count + DRIVE_DIVIDER, a);
  QuickSort (stream, a + DRIVE_DIVIDER, a + b);

  stream->count = a + b;
  stream->match_count = b;

  DBG_LEAVE (); //EXIT //make_stream
}

File: dialogio/pas/make_stream.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2025 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure make_stream(path,mask: String; var stream: tSTREAM);

var
  search: SearchRec;
  count1,count2: Word;
  drive: Char;

{$IFNDEF GO32V2}

{$I ../sdl/CompareStr.pas}

{$ENDIF}

{$I QuickSort.pas}

begin
  _dbg_enter ({$I %FILE%}, 'make_stream');

{$IFNDEF GO32V2}
{$IFNDEF UNIX}
  GetLogicalDriveStrings(SizeOf(drive_list),drive_list);
{$ENDIF}
{$ENDIF}

  count1 := 0;
  For drive := 'A' to 'Z' do
{$IFDEF GO32V2}
    If valid_drive(drive) then
{$ELSE}
    If valid_drive(drive,stream.stuff[SUCC(count1)].info) then
{$ENDIF}
      begin
        Inc(count1);
        stream.stuff[count1].name := drive;
        stream.stuff[count1].attr := volumeid;
        stream.stuff[count1].size := 0;
      end;

  Inc(count1);
  stream.stuff[count1].name := '~'+#255+'~';
  stream.stuff[count1].attr := volumeid;

  count2 := 0;
  stream.drive_count := count1;

{$IFDEF GO32V2}
  If (DiskSize(ORD(UpCase(path[1]))-ORD('A')+1) > 0) then
    begin
{$ENDIF}
      FindFirst(path+WILDCARD_ASTERISK,anyfile-volumeid,search);
      While (DOSerror = 0) and (count1 < MAX_FILES) do
        begin
          If (search.attr AND directory <> 0) and (search.name = '.') then
            begin
              FindNext(search);
              CONTINUE;
            end
          else If (search.attr AND directory <> 0) and
                  NOT ((search.name = '..') and (Length(path) = 3)) then
                 begin
                   If (search.name <> '..') then search.name := search.name
                   else search.name := UPDIR_STR;
                   Inc(count1);
                   stream.stuff[count1].name := search.name;
                   stream.stuff[count1].attr := search.attr;
                 end;
          FindNext(search);
        end;

{$IFNDEF GO32V2}
  If (Length(path) > 3) and (count1 = stream.drive_count) then
    begin
      Inc(count1);
      stream.stuff[count1].name := UPDIR_STR;
      stream.stuff[count1].attr := search.attr;
    end;
{$ENDIF}

      FindFirst(path+WILDCARD_ASTERISK,anyfile-volumeid-directory,search);
      While (DOSerror = 0) and (count1+count2 < MAX_FILES) do
        begin
          If LookUpMask(search.name) then
            begin
          search.name := Lower_filename(search.name);
              Inc(count2);
              stream.stuff[count1+count2].name := search.name;
              stream.stuff[count1+count2].attr := search.attr;
              stream.stuff[count1+count2].size := search.size;
            end;
          FindNext(search);
        end;
{$IFDEF GO32V2}
    end;
{$ENDIF}

  QuickSort(stream.drive_count+DRIVE_DIVIDER,count1);
  QuickSort(count1+DRIVE_DIVIDER,count1+count2);

  stream.count := count1+count2;
  stream.match_count := count2;

  _dbg_leave; //EXIT //make_stream
end;
